docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+2. ___TF -> Swift:___ `Tensor` objects allow for various ways to obtain their content, or a reference to them. One proposal could be to have methods like `.scalar`, `.toArray`, `.iterator`, and `.handle`, that return Swift-typed values. Whenever these are used, data needs to be transferred from TF to Swift and so we would add a `ReceiveFromTF` op to the graph. This actually again consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Enqueue` op will be added to the TF graph, enqueueing the corresponding TF tensor (or its handle, for example) to the queue. At the same time, a `session.run(...)` call will be added to the Swift side, dequeueing the TF content that is to be received from the graph. These method calls will be required to be made explicitly from the library user and will not be added automatically by the Swift compiler, in order to avoid confusion and potential inefficiencies.
+
+## Tacking Point 3: Handling Functions that use TF
+
+Currently, Swift for TF handles all Swift functions that use TF by aggressively inlining them. This can cause lots of issues, including an exploding size of the compiled code, as well as limited supported for recursion (note that TF graph functions themselves support recursion and thus we should be able to support it on our side too). Here we suggest we use TF graph functions to handle Swift functions that use TF. We propose the following high-level design: all Swift functions that use TF will be compiled into two separate functions:
+
+1. ___Swift Function:___ The Swift code part of the compiled function that contains all computation that needs to happen on the Swift side. Note that in the case that this is a pure TF function (using only TF ops), then this function could be compiled to a no-op, or optimized away entirely.
+2. ___TF Graph Function:___ The TF part of the compiled function that is a TF `FunctionDef` protobuf message which contains the extracted TF graph for the compiled function.
+
+The calling code will be invoking both functions in parallel, in its two generated programs -- the Swift program and the TF program, respectively. Note that for the TF program, the invocation will consist of simply adding the corresponding `FunctionDef` to the graph function library and creating an op representing that function call.
+
+Here we omit details related to handling mixed Swift and TF arguments and also related to propagation such function calls which means that one function invocation may require multiple `FunctionDef`s to be added to the graph, as these can be handled relatively easily, and would distract from the main proposed idea.
```

Comment:
That's pretty much what I had in mind. We can also be more conservative about the SIL function type and let it be the full original function type (note that the tensor arguments they could be used in `ReceiveFromTF` calls in the Swift-side code of the function). In this case, if the tensor arguments are not used in the Swift-side code of the function, they will just be ignored and, for now, only incur a small cost of passing them in the stack (if that's not optimized away further down in the compilation process). Although, as you and Alex mention we could also just keep track of what's used where and remove arguments from either signature, as necessary.

In either case, I believe it would be a good idea to standardize this procedure and document it clearly (maybe in this document).